armsrc/lfops.c

```diff
@@ -1286,43 +1355,132 @@ void T55xxResetRead(void) {
+	{
+		case T55xx_DLMode_Fixed    : Timing = &T55xx_Timing_FixedBit;  break; 
+		case T55xx_DLMode_LLR      : Timing = &T55xx_Timing_LLR;	   break;
+		case T55xx_DLMode_Leading0 : Timing = &T55xx_Timing_Leading0;  break;
+		case T55xx_DLMode_1of4     : Timing = &T55xx_Timing_1of4;      break;
+		default:
+				Timing = &T55xx_Timing_FixedBit;
+	}	
+
+	// Build Bit Stream to send.
+	memset (BitStream,0x00,sizeof(BitStream));
+	
+	BitStreamLen = 0;
+	
+	// Add Leading 0 and 1 of 4 reference bit
+	if ((downlink_mode == T55xx_DLMode_Leading0) || (downlink_mode == T55xx_DLMode_1of4))  
+		BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, 0, 1,sizeof(BitStream)); 
+
+	// Add extra reference 0 for 1 of 4
+	if (downlink_mode == T55xx_DLMode_1of4)                            
+		BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, 0, 1,sizeof(BitStream)); 
+
+	// Add Opcode
+	if (testMode) Dbprintf("TestMODE");
+	BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen,testMode ? 0 : 1    , 1,sizeof(BitStream)); 	
+	BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen,testMode ? 1 : Page , 1,sizeof(BitStream)); 	
+	
+	if (PwdMode) {
+
+		// Leading 0 and 1 of 4 00 fixed bits if passsword used		
+		if ((downlink_mode == T55xx_DLMode_Leading0) || (downlink_mode == T55xx_DLMode_1of4)) {
+			BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, 0, 1,sizeof(BitStream)); 
+			BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, 0, 1,sizeof(BitStream)); 
+		}
+		BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, Pwd, 32,sizeof(BitStream)); 
+
+	}
+	// Add Lock bit
+	BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, 0, 1,sizeof(BitStream)); 
 
+	// Add Data if a write command
+	if (!read_cmd)	BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, Data, 32,sizeof(BitStream)); 
+
+	// Add Address
+	if (!reg_readmode) BitStreamLen = T55xx_SetBits (BitStream, BitStreamLen, Block, 3,sizeof(BitStream)); 
+	
+
+
+	// Send Bits to T55xx
+	
 	// Set up FPGA, 125kHz
 	LFSetupFPGAForADC(95, true);
 	StartTicks();
 	// make sure tag is fully powered up...
 	WaitMS(5);
 	// Trigger T55x7 in mode.
 	FpgaWriteConfWord(FPGA_MAJOR_MODE_OFF);
-	WaitUS(START_GAP);
+	WaitUS(Timing->START_GAP);
 
-	if (testMode) Dbprintf("TestMODE");
-	// Std Opcode 10
-	T55xxWriteBit(testMode ? 0 : 1);
-	T55xxWriteBit(testMode ? 1 : Page); //Page 0
 
-	if (PwdMode) {
-		// Send Pwd
-		for (i = 0x80000000; i != 0; i >>= 1)
-			T55xxWriteBit(Pwd & i);
-	}
-	// Send Lock bit
-	T55xxWriteBit(0);
+	// If long leading 0 send long reference pulse
+	if (downlink_mode ==  T55xx_DLMode_LLR) 
+		T55xxWriteBit (T55xx_LongLeadingReference,Timing); // Send Long Leading Start Reference
 
-	// Send Data
-	for (i = 0x80000000; i != 0; i >>= 1)
-		T55xxWriteBit(Data & i);
+	
+	if (downlink_mode ==  T55xx_DLMode_1of4) { // 1 of 4 need to send 2 bits at a time
+		for (i = 0; i < BitStreamLen; i+=2) {
+			SendBits = (BitStream[i] << 1) + BitStream[i+1]; 
+			T55xxWriteBit (SendBits,Timing);
+		}
```

Comment:
My view was that we built the BitStream as called so the client could not affect the bit count.  The buffer was bigger then needed, and set to 0.  so 1 bit over, if it did happen would then clock out 10.
Given that 1 of 4 MUST have two bits at a time, then I cant see how it could go over.
e.g. 8 bits thus 0 - 7  0,2,4,6 (8 >= 7) then the +1  would 1,3,5,7
Longest Packet would be the 1 of 4 write with password.
00 (ref) 10 (op) 00 (fixed) (32 bit password) L (32 bit data) (3 bit addr)  74 bits
So if we stick with the bool, then drop that down to 76 (just in case)
If we convert back to the other bitstream , then 10 bytes (80bits).

Happy to set what you want.

Reply:
you are correct that the existing code has the extra memory allocated to the BitStream already and it should currently always be even. 
i was just thinking for future editors, i believe it would be safer to write
`for (i = 0; i < BitStreamLen-1; i+=2) {`

Reply:
> Now just the reset test ?
the resetread is a way of ensuring you capture the start of the data transmission. as the tag will be powered up and then be forced to restart it's stream (upon successful reset read command)
this removes the power up delays or other normal lf read capture delays.
multiple lf t55xx resetread s should always output the same trace (within a small margin of error, esp. if tag doesn't move..).  (multiple lf read s will not be as precise)

Reply:
I have added the i < BitStreamLen -1;
As is (with change), BitstreamLen and Bitstream Array are local to the function with the loop to clock out the bits.
BitStreamLen is an uint8_t, so cant be < 0.  If <= 0 then the loop wont action (so should be covered)
i.e. for ( i = 0; i < BitStreamLen-1; i+=2 )  
and for (i = 0; i < BitStreamLen; i++) 
0 cant  be less then 0.  
So now with the -1, BitStreamLen = 0 wont loop.  BitStreamLen = 1 wont loop in the 1 of 4 but will clock out a single bit for others, > 1 should work as designed.
(will push after testing again)

Reply:
Thanks for the reset test help.  It did show it was not working as needed.  fixed now and test shows 0-1 pixel shift in the data plot (most of the time 0 change), so as you explained.

Reply:
Retested all lf t55 commands (in all downlink modes where supported) with and without password.  
lf t55 reset - tested and working as explained.
lf hid clone and lf em 410xwrite tested
Fork updated.

Reply:
If you subtract 1 from 0 for a uint8_t you get 255.  See
https://stackoverflow.com/questions/7221409/is-unsigned-integer-subtraction-defined-behavior

Reply:
Yes, understand.
Since, as is, we have a max 80 bit bit-stream (10 bytes), I can change the unit8_t to an int8_t (-128 to 128) i.e. int8_t BitStreamLen; (and update the loop i to int8_t so as to not be comparing a unit to int.
That should fix with minimal overhead.  

Another option would be a check if > 0, a little more overhead,   but lets us grow the buffer to > if ever needed
i.e.  if ((downlink_mode ==  T55xx_DLMode_1of4) && ( BitStreamLen > 0))

Just let me know which way you would like to go and I will change as needed.

Reply:
I'd just add the condition to the if statement.

Reply:
Done.